controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -339,11 +339,41 @@ func (r *BareMetalHostReconciler) credentialsErrorResult(err error, request ctrl
 }
 
 // hasRebootAnnotation checks for existence of reboot annotations and returns true if at least one exist
-func hasRebootAnnotation(host *metal3v1alpha1.BareMetalHost) bool {
-	for annotation := range host.Annotations {
+func hasRebootAnnotation(info *reconcileInfo) (hasReboot, hardMode bool) {
+	for annotation, value := range info.host.GetAnnotations() {
+		// Don't use a break here as we may have multiple clients setting
+		// reboot annotations and we always want hard requests honoured
 		if isRebootAnnotation(annotation) {
-			return true
+			hasReboot = true
+			if isHardReboot(value, info) {
+				hardMode = true
+			}
+		}
+	}
+	return
+}
+
+// isHardReboot checks if reboot annotation requests a hard reboot and returns true if so
+func isHardReboot(annotation string, info *reconcileInfo) bool {
+
+	type Annotations struct {
+		Mode	string	`json:"mode"`
+		Value	string	`json:"value"`
+	}
+
+	annotations := Annotations{}
+	err := json.Unmarshal([]byte(annotation), &annotations)
+	if err != nil {
+		if annotation != "" {
+			info.publishEvent("InvalidAnnotationValue", fmt.Sprintf("could not parse reboot annotation (%s) - invalid json, assuming soft-reboot", annotation))
+			info.log.Info(fmt.Sprintf("Could not parse reboot annotation ('%s') - invalid json, assuming soft-reboot", annotation))
+		} else {
+			info.log.Info("No reboot annotation value specified, assuming soft-reboot.")
 		}
+		return false
+	}
+	if annotations.Mode == "hard" {
```

Comment:
Instead of checking the string literal, this should check a const defined in the types file with the struct.

Reply:
Sounds good.